Sources/PhishingDetection/PhishingDetectionDataStore.swift

```diff
@@ -0,0 +1,134 @@
+	public var hashValue: String
+	public var regex: String
+
+	enum CodingKeys: String, CodingKey {
+		case hashValue = "hash"
+		case regex
+	}
+
+	public init(hashValue: String, regex: String) {
+		self.hashValue = hashValue
+		self.regex = regex
+	}
+}
+
+public struct Match: Decodable, Encodable, Hashable {
+	var hostname: String
+	var url: String
+	var regex: String
+	var hash: String
+
+	public init(hostname: String, url: String, regex: String, hash: String) {
+		self.hostname = hostname
+		self.url = url
+		self.regex = regex
+		self.hash = hash
+	}
+}
+
+public protocol PhishingDetectionDataStoring {
+    var filterSet: Set<Filter> { get set }
+    var hashPrefixes: Set<String> { get set }
+    var currentRevision: Int { get set }
+    func writeData()
+    func loadData() async
+}
+
+public class PhishingDetectionDataStore: PhishingDetectionDataStoring {
+    public var filterSet: Set<Filter> = []
+    public var hashPrefixes = Set<String>()
+    public var currentRevision = 0
+
+    var dataProvider: PhishingDetectionDataProviding
+    var dataStore: URL?
+    var hashPrefixFilename: String = "hashPrefixes.json"
+    var filterSetFilename: String = "filterSet.json"
+    var revisionFilename: String = "revision.txt"
+
+    public init(dataProvider: PhishingDetectionDataProviding) {
+        self.dataProvider = dataProvider
+        createFileDataStore()
+    }
+
+    func createFileDataStore() {
+        do {
+            let fileManager = FileManager.default
+            let appSupportDirectory = try fileManager.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
+            dataStore = appSupportDirectory.appendingPathComponent(Bundle.main.bundleIdentifier!, isDirectory: true)
+            try fileManager.createDirectory(at: dataStore!, withIntermediateDirectories: true, attributes: nil)
+        } catch {
+            os_log(.debug, log: .phishingDetection, "\(self): 🔴 Error creating phishing protection data directory: \(error)")
+        }
+    }
+
+    public func writeData() {
+        let encoder = JSONEncoder()
+        do {
+            let hashPrefixesData = try encoder.encode(Array(hashPrefixes))
+            let filterSetData = try encoder.encode(Array(filterSet))
+            let revision = try encoder.encode(self.currentRevision)
+
+            let hashPrefixesFileURL = dataStore!.appendingPathComponent(hashPrefixFilename)
+            let filterSetFileURL = dataStore!.appendingPathComponent(filterSetFilename)
+            let revisionFileURL = dataStore!.appendingPathComponent(revisionFilename)
+
+            try hashPrefixesData.write(to: hashPrefixesFileURL)
+            try filterSetData.write(to: filterSetFileURL)
+            try revision.write(to: revisionFileURL)
+        } catch {
+            os_log(.debug, log: .phishingDetection, "\(self): 🔴 Error saving phishing protection data: \(error)")
+        }
+    }
+
+    public func loadData() async {
```

Comment:
Just some ideas:

maybe on loadData we could just write the embedded data to the file URL if none available and we could make the variables lazy variable
Something like this:

   lazy var filterSet: Set<Filter> = {
        loadSet(from: "filterSet.json") ?? dataProvider.loadEmbeddedFilterSet()
    }() 
    
    (and then you update them from the write func